### PR TITLE
Fix/issue 9 Transaction for start & end experiment

### PR DESCRIPTION
**Pull request by RidhamShah**
In this PR, we have created a transaction for the start experiment and end experiment of the scheduler. Along with it, we are throwing an error if the time window, time at function is called and the timeStamp in Scheduler Repository is more than 5 hours.

### CODE REVIEW

File: backend/packages/Upgrade/src/api/repositories/ExperimentAuditLogRepository.ts [hunk 1]
```diff
@@ -35,9 +35,17 @@ export class ExperimentAuditLogRepository extends Repository<ExperimentAuditLog>
...
 
-  public async saveRawJson(type: EXPERIMENT_LOG_TYPE, data: any, user: User): Promise<ExperimentAuditLog> {
-    const result = await this.createQueryBuilder()
+  public async saveRawJson(
+    type: EXPERIMENT_LOG_TYPE,
+    data: any,
+    user: User,
+    entityManger?: EntityManager
+  ): Promise<ExperimentAuditLog> {
+    const that = entityManger ? entityManger : this;
```
Comment (JD2455) [inline backend/packages/Upgrade/src/api/repositories/ExperimentAuditLogRepository.ts:44]: I think we can update this later by OR operator also.
`const that = entityManager || this`

File: backend/packages/Upgrade/src/api/repositories/ExperimentRepository.ts [hunk 2]
```diff
@@ -84,10 +84,13 @@ export class ExperimentRepository extends Repository<Experiment> {
     state: EXPERIMENT_STATE,
     scheduleDate: Date,
     endDate: Date = null,
-    startDate: Date = null
+    startDate: Date = null,
+    entityManager?: EntityManager
   ): Promise<Experiment> {
-    const result = await this.createQueryBuilder('experiment')
-      .update()
+    const that = entityManager ? entityManager : this;
```
Comment (JD2455) [inline backend/packages/Upgrade/src/api/repositories/ExperimentRepository.ts:90]: Same as above

File: backend/packages/Upgrade/src/api/services/ScheduledJobService.ts [hunk 3]
```diff
@@ -6,57 +6,92 @@ import { ScheduledJob, SCHEDULE_TYPE } from '../models/ScheduledJob';
...
+        const fiveHoursInMS = 18000000;
+        
+        if (timeDif > fiveHoursInMS) {
+          const errorMsg =  'Time Differnce of more than 5 hours is found';
+          await transactionalEntityManager.getRepository(ScheduledJob).delete({id: scheduledJob.id});
+          throw new Error(errorMsg);
+        }
+
+        if (scheduledJob && scheduledJob.experiment) {
+          const experiment = await transactionalEntityManager.getRepository(Experiment).findOne(scheduledJob.experiment.id);
```
Comment (JD2455) [inline backend/packages/Upgrade/src/api/services/ScheduledJobService.ts:43]: We can extract the Experiment repository only once and then use the same instead of getting every time from the transactionalEntityManager.

**Review by VivekFitkariwala — APPROVED** [review]

File: backend/packages/Upgrade/src/api/services/ScheduledJobService.ts [hunk 4]
```diff
@@ -6,57 +6,92 @@ import { ScheduledJob, SCHEDULE_TYPE } from '../models/ScheduledJob';
...
-      return experimentService.updateState(
-        scheduledJob.experiment.id,
-        EXPERIMENT_STATE.ENROLLMENT_COMPLETE,
-        systemUser
-      );
-    }
-    return {};
+    return await getConnection().transaction(async (transactionalEntityManager) => {
+      try {
+        const scheduledJob = await transactionalEntityManager.getRepository(ScheduledJob).findOne(id, { relations: ['experiment'] });
```
Comment (JD2455) [inline backend/packages/Upgrade/src/api/services/ScheduledJobService.ts:63]: We can extract the required repository only once and then use the same instead of getting every time from the transactionalEntityManager.

File: backend/packages/Upgrade/src/api/repositories/ExperimentAuditLogRepository.ts (hunk 1 above)
Comment (RidhamShah) [reply backend/packages/Upgrade/src/api/repositories/ExperimentAuditLogRepository.ts]: Code Updated

File: backend/packages/Upgrade/src/api/repositories/ExperimentRepository.ts (hunk 2 above)
Comment (RidhamShah) [reply backend/packages/Upgrade/src/api/repositories/ExperimentRepository.ts]: Code Updated

File: backend/packages/Upgrade/src/api/services/ScheduledJobService.ts (hunk 3 above)
Comment (RidhamShah) [reply backend/packages/Upgrade/src/api/services/ScheduledJobService.ts]: Code Updated

File: backend/packages/Upgrade/src/api/services/ScheduledJobService.ts (hunk 4 above)
Comment (RidhamShah) [reply backend/packages/Upgrade/src/api/services/ScheduledJobService.ts]: Code Updated

File: backend/packages/Upgrade/src/api/repositories/ExperimentAuditLogRepository.ts [hunk 5]
```diff
@@ -1,28 +1,29 @@
...
-        .where('audit.type = :filter', { filter });
+      queryBuilder = queryBuilder.where('audit.type = :filter', { filter });
     }
-    return queryBuilder
-      .getMany()
-      .catch((error: any) => {
-        const errorMsg = repositoryError('ExperimentAuditLogRepository', 'paginatedFind', { limit, offset }, error);
-        throw new Error(errorMsg);
-      });
+    return queryBuilder.getMany().catch((error: any) => {
```
Comment (JD2455) [inline backend/packages/Upgrade/src/api/repositories/ExperimentAuditLogRepository.ts:23]: @RidhamShah No new code changes here but formatting is updated. Can you please revert the changes?

File: backend/packages/Upgrade/src/api/services/ScheduledJobService.ts [hunk 6]
```diff
@@ -6,57 +6,106 @@ import { ScheduledJob, SCHEDULE_TYPE } from '../models/ScheduledJob';
...
-        // update experiment startOn
-        await this.experimentRepository.update({ id: experiment.id }, { startOn: null });
-        return experimentService.updateState(scheduledJob.experiment.id, EXPERIMENT_STATE.ENROLLING, systemUser);
+    return await getConnection().transaction(async (transactionalEntityManager) => {
+      try {
+        const scheduledJobRepository = transactionalEntityManager.getRepository(ScheduledJob);
+        const scheduledJob = await scheduledJobRepository.findOne(id, { relations: ['experiment'] });
+
+        const currentDate = new Date();
+        const timeDif = Math.abs(currentDate.getTime() - scheduledJob.timeStamp.getTime());
```
Comment (JD2455) [inline backend/packages/Upgrade/src/api/services/ScheduledJobService.ts:34]: @RidhamShah Update variable name to timeDiff

File: backend/packages/Upgrade/src/api/services/ScheduledJobService.ts [hunk 7]
```diff
@@ -6,57 +6,106 @@ import { ScheduledJob, SCHEDULE_TYPE } from '../models/ScheduledJob';
...
-    return {};
+    return await getConnection().transaction(async (transactionalEntityManager) => {
+      try {
+        const schedulerJobRepository = transactionalEntityManager.getRepository(ScheduledJob);
+        const scheduledJob = await schedulerJobRepository.findOne(id, { relations: ['experiment'] });
+        const experimentRepository = transactionalEntityManager.getRepository(Experiment);
+        const experiment = await experimentRepository.findOne(scheduledJob.experiment.id);
+
+        const currentDate = new Date();
+        const timeDif = Math.abs(currentDate.getTime() - scheduledJob.timeStamp.getTime());
```
Comment (JD2455) [inline backend/packages/Upgrade/src/api/services/ScheduledJobService.ts:79]: @RidhamShah Update variable name to timeDiff

File: backend/packages/Upgrade/src/api/services/ScheduledJobService.ts [hunk 8]
```diff
@@ -6,57 +6,106 @@ import { ScheduledJob, SCHEDULE_TYPE } from '../models/ScheduledJob';
...
-      return experimentService.updateState(
-        scheduledJob.experiment.id,
-        EXPERIMENT_STATE.ENROLLMENT_COMPLETE,
-        systemUser
-      );
-    }
-    return {};
+    return await getConnection().transaction(async (transactionalEntityManager) => {
+      try {
+        const schedulerJobRepository = transactionalEntityManager.getRepository(ScheduledJob);
```
Comment (JD2455) [inline backend/packages/Upgrade/src/api/services/ScheduledJobService.ts:73]: @RidhamShah Can you please update the name to scheduledJobRepository instead of schedulerJobRepository?

Comment (YashilDepani) [reply backend/packages/Upgrade/src/api/services/ScheduledJobService.ts:73]: @RidhamShah if done then let us know.

File: backend/packages/Upgrade/src/api/services/ScheduledJobService.ts (hunk 6 above)
Comment (RidhamShah) [reply backend/packages/Upgrade/src/api/services/ScheduledJobService.ts]: Code Updated

File: backend/packages/Upgrade/src/api/services/ScheduledJobService.ts (hunk 7 above)
Comment (RidhamShah) [reply backend/packages/Upgrade/src/api/services/ScheduledJobService.ts]: Code Updated

File: backend/packages/Upgrade/src/api/services/ScheduledJobService.ts (hunk 8 above)
Comment (RidhamShah) [reply backend/packages/Upgrade/src/api/services/ScheduledJobService.ts]: Code Updated

**Review by JD2455 — APPROVED** [review]
